# T05- Motor Trend Car Road Test

1.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Considera todos los demás factores como numéricos/ordinales.

Calcula el R2 e interpreta los signos de los betas.
Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.
Añade regularización L2 con un hiperparámetro lambda decidido por ti. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba.

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.metrics import r2_score
import statsmodels.api as sm
from sklearn.metrics import mean_squared_error

In [7]:
df = pd.read_excel("Motor Trend Car Road Tests.xlsx")
df= df.drop(columns="model")

In [8]:
x=df.drop(columns="mpg")
y=df["mpg"]


1.1

In [9]:
modelo=LinearRegression()
modelo.fit(x,y)
y_pred=modelo.predict(x)
r2=r2_score(y,y_pred)
print(r2,modelo.coef_,modelo.intercept_)

0.8690157644777647 [-0.11144048  0.01333524 -0.02148212  0.78711097 -3.71530393  0.82104075
  0.31776281  2.52022689  0.65541302 -0.19941925] 12.30337415599622


*   El modelo explica casi el 87% de la variabilidad.
cyl (-0.111) → más cilindros reducen ligeramente el mpg (coherente: motores con más cilindros consumen más).

disp (+0.013) → mayor desplazamiento aumenta levemente el mpg, lo cual es contraintuitivo; seguramente se debe a multicolinealidad con cyl y hp.

hp (-0.021) → más caballos de fuerza reducen el mpg (esperado: motores más potentes gastan más).

drat (+0.787) → una relación del diferencial más alta se asocia con mayor mpg (coherente con transmisiones más eficientes).

wt (-3.715) → el peso tiene un fuerte efecto negativo: autos más pesados consumen más.

qsec (+0.821) → un mayor tiempo en el cuarto de milla (autos más lentos) se relaciona con más mpg (tiene sentido: autos menos deportivos son más eficientes).

vs (+0.318) → motores en línea tienden a consumir menos que los motores en V.

am (+2.520) → transmisión manual se asocia con más mpg frente a automática (en esta época los manuales eran más eficientes).

gear (+0.655) → más marchas aumentan el mpg (mejor aprovechamiento del motor).

carb (-0.199) → más carburadores reducen el mpg (esperado: motores con más carburadores son más tragones).




In [10]:
Xtrain, Xtest, ytrain, ytest = train_test_split(x, y, train_size=0.4,random_state=137)
#modelo = LinearRegression()
#modelo.fit(Xtrain, ytrain)
#y_train_pred = modelo.predict(Xtrain)
#y_test_pred = modelo.predict(Xtest)
#r2_score(ytest, y_test_pred), r2_score(ytrain, y_train_pred)

(-54.0717852041882, 0.9900805114414539)

Podemos observar que es un muy mal modelo.

# L2
LAMBDA = 1

In [11]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler().fit(Xtrain)
Xtrain = scaler.transform(Xtrain)
Xtest = scaler.transform(Xtest)

In [12]:
ridge = Ridge(alpha=1.0)
ridge.fit(Xtrain, ytrain)
y_train_ridge = ridge.predict(Xtrain)
y_test_ridge = ridge.predict(Xtest)
r2_score(ytest, y_test_ridge), r2_score(ytrain, y_train_ridge)

(0.5910911509384542, 0.9386746881405903)

LAMBDA = 0.5

In [13]:
ridge = Ridge(alpha=0.5)
ridge.fit(Xtrain, ytrain)
y_train_ridge = ridge.predict(Xtrain)
y_test_ridge = ridge.predict(Xtest)
r2_score(ytest, y_test_ridge), r2_score(ytrain, y_train_ridge)

(0.4898013499696513, 0.9579016413911189)

LAMBDA = 2

In [14]:
ridge = Ridge(alpha=2.0)
ridge.fit(Xtrain, ytrain)
y_train_ridge = ridge.predict(Xtrain)
y_test_ridge = ridge.predict(Xtest)
r2_score(ytest, y_test_ridge), r2_score(ytrain, y_train_ridge)

(0.6660803311091932, 0.917533101368156)

LAMBDA = 3

In [15]:
ridge = Ridge(alpha=3.0)
ridge.fit(Xtrain, ytrain)
y_train_ridge = ridge.predict(Xtrain)
y_test_ridge = ridge.predict(Xtest)
r2_score(ytest, y_test_ridge), r2_score(ytrain, y_train_ridge)

(0.699343946116874, 0.9053841657562541)

Con un lambda grande nos da un mejor R2 test.

# 1.2

In [16]:
x1=df.drop(columns="qsec")
y1=df["qsec"]

In [17]:
modelo1=LinearRegression()
modelo1.fit(x1,y1)
y_pred1=modelo1.predict(x1)
r2_1=r2_score(y1,y_pred1)
print(r2_1,modelo1.coef_,modelo1.intercept_)

0.8746925777094 [ 0.06904768 -0.36267823 -0.00750084 -0.00156255 -0.1310636   1.49633224
  0.97003474 -0.90118641 -0.20128511 -0.27359751] 17.77617692970135


*   El modelo explica 87.5% de la variabilidad.

mpg (+0.069) → a mayor eficiencia de combustible, el auto tiende a tardar más en el ¼ de milla. Tiene sentido: autos “económicos” suelen ser menos deportivos y por lo tanto más lentos.

cyl (−0.363) → más cilindros reducen el tiempo en el ¼ de milla. Autos con más cilindros suelen tener más potencia y aceleran más rápido.

disp (−0.0075) → mayor desplazamiento también reduce el tiempo, coherente con motores más grandes y rápidos.

hp (−0.0016) → más caballos de fuerza reducen el tiempo, lo esperado (más potencia = mayor aceleración).

drat (−0.131) → una relación de diferencial mayor se asocia con tiempos más bajos; puede estar reflejando que ciertas configuraciones de transmisión permiten acelerar más rápido.

wt (+1.496) → autos más pesados tardan bastante más en el ¼ de milla; este es uno de los efectos más fuertes y tiene mucho sentido: más masa = menos aceleración.

vs (+0.970) → motores en línea (vs=1) se asocian con un tiempo mayor que motores en V; los V suelen ser más deportivos.

am (−0.901) → transmisión manual reduce el tiempo en el ¼ de milla, es decir, los autos manuales son más rápidos en aceleración que los automáticos en este dataset.

gear (−0.201) → más marchas reducen un poco el tiempo, lo que encaja con transmisiones optimizadas para performance.

carb (−0.274) → más carburadores reducen el tiempo, consistente con motores más potentes y deportivos.



In [29]:
Xtrain1, Xtest1, ytrain1, ytest1 = train_test_split(x1, y1, train_size=0.4,random_state=137)
#modelo1 = LinearRegression()
#modelo1.fit(Xtrain1, ytrain1)
#y_train_pred1 = modelo1.predict(Xtrain1)
#y_test_pred1 = modelo1.predict(Xtest1)
#r2_score(ytest1, y_test_pred1), r2_score(ytrain1, y_train_pred1)

#L2 para qsec

LAMBDA= 1

In [38]:
from sklearn.preprocessing import StandardScaler
Xtrain1, Xtest1, ytrain1, ytest1 = train_test_split(x1, y1, train_size=0.4,random_state=137)
scaler = StandardScaler().fit(Xtrain1)

Xtrain_scaled = scaler.transform(Xtrain1)
Xtest_scaled = scaler.transform(Xtest1)

In [39]:
rd = Ridge(alpha=1)
rd.fit(Xtrain_scaled, ytrain1)
y_train_pred = rd.predict(Xtrain_scaled)
y_test_pred = rd.predict(Xtest_scaled)

r2_score(ytest1, y_test_pred), r2_score(ytrain1, y_train_pred)

(0.6535717344645446, 0.9671802657602482)

LAMBDA=0.5

In [40]:
rd = Ridge(alpha=0.5)
rd.fit(Xtrain_scaled, ytrain1)
y_train_pred = rd.predict(Xtrain_scaled)
y_test_pred = rd.predict(Xtest_scaled)

r2_score(ytest1, y_test_pred), r2_score(ytrain1, y_train_pred)

(0.6424218216092722, 0.9727495500263381)

LAMBDA=2

In [41]:
rd = Ridge(alpha=2)
rd.fit(Xtrain_scaled, ytrain1)
y_train_pred = rd.predict(Xtrain_scaled)
y_test_pred = rd.predict(Xtest_scaled)

r2_score(ytest1, y_test_pred), r2_score(ytrain1, y_train_pred)

(0.6575165192963932, 0.9562802655200474)

# 2.1

In [42]:
x_dummies = pd.get_dummies(x, columns=["cyl", "gear", "carb"], drop_first=True)
modelo=LinearRegression()
modelo.fit(x_dummies,y)
y_pred=modelo.predict(x_dummies)
r2=r2_score(y,y_pred)
print(r2,modelo.coef_,modelo.intercept_)

0.8930749320864843 [ 0.03554632 -0.07050683  1.18283018 -4.52977584  0.36784482  1.93085054
  1.2121157  -2.64869528 -0.33616298  1.11435494  2.52839599 -0.97935432
  2.99963875  1.09142288  4.47756921  7.25041126] 23.87913244333949


*   El modelo explica casi 89.3% de la variabilidad en mpg.
disp (0.0355) → más desplazamiento se asocia con un leve aumento de mpg (posiblemente efecto de correlación con otras variables).

hp (−0.0705) → más potencia reduce el mpg, lo esperado.

drat (+1.1828) → relaciones de diferencial más altas mejoran la eficiencia.

wt (−4.5298) → el peso reduce fuertemente el mpg (autos más pesados consumen más).

qsec (+0.3678) → autos con mayor tiempo en ¼ de milla (más lentos) tienden a tener más mpg (coherente: menos deportivos = más eficientes).

vs (+1.9309) → motores en línea consumen menos que los motores en V.

am (+1.2121) → transmisión manual mejora el mpg respecto a automática.

dummies de cyl (ej. −2.6487, −0.3362, etc.) → comparadas con la categoría base (p. ej. 4 cilindros), los autos de 6 u 8 cilindros presentan menor mpg (lo esperado).

dummies de gear (ej. +1.1144, +2.5284, etc.) → comparados con la categoría base (p. ej. 3 marchas), más marchas suelen asociarse a mayor mpg (lógico: mejor aprovechamiento de motor).

dummies de carb (ej. −0.9794, +2.9996, +1.0914, +4.4776, +7.2504) → según el número de carburadores respecto a la categoría base, algunos aumentan y otros disminuyen mpg. En general, más carburadores suelen reducir la eficiencia, pero aquí aparecen coeficientes positivos por multicolinealidad y lo pequeño del dataset.



In [24]:
Xtrain_dummies, Xtest_dummies, ytrain_dummies, ytest_dummies = train_test_split(x_dummies, y, train_size=0.4,random_state=137)
modelo_dummies = LinearRegression()
modelo_dummies.fit(Xtrain_dummies, ytrain_dummies)
y_train_pred_dummies = modelo_dummies.predict(Xtrain_dummies)
y_test_pred_dummies = modelo_dummies.predict(Xtest_dummies)
r2_score(ytest_dummies, y_test_pred_dummies), r2_score(ytrain_dummies, y_train_pred_dummies)

(-0.03244324027529921, 1.0)

Nuevamente obtuvimos un terrible modelo.

#2.2

In [25]:
x1_dummies = pd.get_dummies(x1, columns=["cyl", "gear", "carb"], drop_first=True)
modelo=LinearRegression()
modelo.fit(x1_dummies,y1)
y_pred1_dummies=modelo.predict(x1_dummies)
r2=r2_score(y1,y_pred1_dummies)
print(r2,modelo.coef_,modelo.intercept_)

0.9082689440167556 [ 2.77413839e-02  3.53115688e-03 -2.06618843e-03  1.07865814e-01
  8.10472140e-01  2.65732208e-01 -1.69429351e+00 -1.10434255e+00
 -2.96690068e+00  1.33228234e+00  1.82317492e-01 -8.34412978e-01
 -2.29304367e-01 -1.94703381e+00 -1.56524104e+00 -1.33231717e+00] 16.349137498224888


mpg (+0.028): autos más eficientes (más mpg) tienden a tardar más en el ¼ de milla → menos deportivos.

disp (+0.0035): motores con más desplazamiento aumentan un poco el tiempo; contraintuitivo, puede ser efecto de colinealidad.

hp (−0.0021): más caballos reducen el tiempo (coherente: más potencia = más aceleración).

drat (+0.108): diferencial más alto se asocia con tiempos mayores (posible efecto de correlación con marchas/transmisión).

wt (+0.810): autos más pesados tardan más (muy lógico).

vs (+0.266): motores en línea tardan más que motores en V (los V suelen ser más deportivos).

am (−1.694): transmisión manual reduce el tiempo (manuales aceleran mejor que automáticos en este dataset).

cyl_6 (−1.104), cyl_8 (−2.967): comparados con 4 cilindros (categoría base), los autos de 6 y 8 cilindros son más rápidos (menos tiempo en ¼ milla).

gear_4 (+1.332), gear_5 (+0.182): comparados con 3 marchas (categoría base), los autos con 4 o 5 marchas tardan más, lo cual sorprende (podría reflejar la composición del dataset más que una ley general).

carb_2 (−0.834), carb_3 (−0.229), carb_4 (−1.947), carb_6 (−1.565), carb_8 (−1.332): frente a 1 carburador (categoría base), tener más carburadores reduce el tiempo (los autos con más carburadores suelen ser deportivos y aceleran más rápido).

In [26]:
Xtrain_dummies1, Xtest_dummies1, ytrain_dummies1, ytest_dummies1 = train_test_split(
    x1_dummies, y1, train_size=0.4, random_state=137)

modelo_dummies1 = LinearRegression()
modelo_dummies1.fit(Xtrain_dummies1, ytrain_dummies1)
y_train_pred_dummies1 = modelo_dummies1.predict(Xtrain_dummies1)
y_test_pred_dummies1 = modelo_dummies1.predict(Xtest_dummies1)
r2_score(ytest_dummies1, y_test_pred_dummies1),r2_score(ytrain_dummies1, y_train_pred_dummies1)


(-2.1319559728357165, 1.0)

SOBREAJUSTE EXTREMO SOS

# 3.1
Ejercicio 1.1:

R² (todo el dataset) ≈ 0.869.

Con Ridge (L2, λ entre 0.5 y 2), el R² de test mejoró a valores positivos (0.48–0.66), mostrando que la regularización ayuda.

Ejercicio 2.1 (con dummies en cyl, gear, carb):

R² (todo el dataset) ≈ 0.893, ligeramente mejor que 1.1.

Con los el train y test dan cosas horribles.

# 3.2
(y = qsec)

Ejercicio 1.2:

R² (todo el dataset) ≈ 0.875.

Con Ridge (L2), el R² test mejoró hasta 0.65–0.66.

Ejercicio 2.2 (con dummies):

R² (todo el dataset) ≈ 0.908, mejor que 1.2.

Y luego que lo separas en train y test vuelven a dar cosas horribles, ya que estan muy pocos datos